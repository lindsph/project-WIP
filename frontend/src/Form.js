import React, {useState} from 'react';
import { Link } from "react-router-dom";
import { useForm } from 'react-hook-form';
import styles from './Form.module.css';

const Form = ({login, signup}) => {
    const { register, handleSubmit, watch, errors, reset } = useForm({
        defaultValues: {
            email: "Email",
            password: "Password"
        }
    });

    // will watch specified inputs and return their values
    // const watchAllFields = watch();
    // console.log(watchAllFields)

    const onSubmit = async data => {
        // console.log(data);
        // returns an object with email and password as keys and the values entered into the inputs as values

        // need to add logic to either signup or login..
        await login(data.email, data.password);
        reset();
    }

    return (
        <div className={styles['form-wrapper']}>
            <h1 className={styles['form-heading']}>
                {login ? 'Sign in' : 'Create account'}
            </h1>
            {login ? (
                <h2 className={styles['form-heading']}>
                    Don't have an account? <Link to="/sign_up">Create one</Link> here.
                </h2>
            ) : (
                <h2 className={styles['form-heading']}>
                    Already have an account? <Link to="/sign_in">Sign in</Link> instead.
                </h2>
            )}
            {/* handleSubmit will validate your inputs before invoking onSubmit */}
            <form onSubmit={handleSubmit(onSubmit)}>
                {/* register your input into the hook by invoking the register function, must have a name as a key for registration */}
                {/* default value for an input is the initial value when a component is frist rendered */}
                {/* difference between that and value is that value indicates the input from the user */}
                {/* required is a boolean which if true indicates the input must have a value before the form can be submitted */}
                <div className={styles['form-group']}>
                    <label htmlFor="email">Email Address</label>
                    <input
                        type="text"
                        name="email"
                        id="email"
                        aria-invalid={errors.email ? "true" : "false"}
                        ref={register(
                            {
                                required: true,
                                pattern: {
                                    value: /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
                                    message: "Email does not match format"
                                },
                            })}
                    />
                    {errors?.email && <span role="alert" className={styles['notification-message']}>{errors.email.message}</span>}
                </div>
                <div className={styles['form-group']}>
                    <label htmlFor="password">Password</label>
                    <input
                        type="text"
                        name="password"
                        id="password"
                        aria-invalid={errors.password ? "true" : "false"}
                        ref={register(
                            {
                                required: true,
                                minLength: {
                                    value: 5,
                                    message: "Minimum password length is 5"
                                }
                            })}
                    />
                    {errors?.password && <span role="alert" className={styles['notification-message']}>{errors.password.message}</span>}
                </div>

                <button type="submit">Submit</button>
            </form>
        </div>
    );
};

export default Form;