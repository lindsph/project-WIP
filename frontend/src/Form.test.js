import React from "react";
import 'mutationobserver-shim';
import { render, screen, fireEvent, waitFor, getByText, findByText, act } from "@testing-library/react";
import Form from "./Form";

global.MutationObserver = window.MutationObserver;

const mockLogin = jest.fn((email, password) => {
    return { email, password };
});

const createDefaultProps = () => ({
    login: jest.fn(),
    heading: 'heading here'
});

describe("<Form />", () => {

    it('renders correctly with the default props', () => {
        const props = createDefaultProps();

        render(<Form {...props}/>)

        expect(document.body).toMatchSnapshot();
    });

    it("should display required error when value for email is invalid and not call login function", async () => {
        const { getByText } = render(<Form login={mockLogin} />);
        const submitButton = getByText("Submit");

        fireEvent.click(submitButton);

        expect(await screen.findAllByRole("alert")).toHaveLength(1);
        expect(mockLogin).not.toBeCalled();
    });

    it("should display correct error when email is invalid", async () => {
        const { getByText } = render(<Form login={mockLogin} />);
        const submitButton = getByText("Submit");

        fireEvent.input(screen.getByLabelText("Email Address"), {
            target: {
                value: "invalidEmail"
            }
        });

        fireEvent.input(screen.getByLabelText("Password"), {
            target: {
                value: "password"
            }
        });

        fireEvent.click(submitButton);

        expect(await screen.findAllByRole("alert")).toHaveLength(1);
        expect(mockLogin).not.toBeCalled();
        expect(screen.getByLabelText("Email Address").value).toBe("invalidEmail");
        expect(screen.getByLabelText("Password").value).toBe("password");
    });

    it("should display min length error when password is invalid (and email is valid)", async () => {
        const { getByText } = render(<Form login={mockLogin} />);
        const submitButton = getByText("Submit");

        fireEvent.input(screen.getByLabelText("Email Address"), {
            target: {
                value: "test@email.com"
            }
        });

        fireEvent.input(screen.getByLabelText("Password"), {
            target: {
                value: "pass"
            }
        });

        fireEvent.click(submitButton);

        expect(await screen.findAllByRole("alert")).toHaveLength(1);
        expect(mockLogin).not.toBeCalled();
        expect(screen.getByLabelText("Password").value).toBe("pass");
    });

    it("should not display error when values are valid", async () => {
        const { getByText } = render(<Form login={mockLogin} />);
        const submitButton = getByText("Submit");

        fireEvent.input(screen.getByLabelText("Email Address"), {
            target: {
                value: "test@email.com"
            }
        });

        fireEvent.input(screen.getByLabelText("Password"), {
            target: {
                value: "password"
            }
        });

        fireEvent.click(submitButton);

        const result = mockLogin("test@email.com", "password");

        expect(result).toStrictEqual({email: "test@email.com", password: "password"});
        expect(mockLogin).toHaveBeenCalled();
        expect(mockLogin).toHaveBeenCalledWith("test@email.com", "password")
        // expect(screen.getByLabelText("Password").value).toBe("");
        // expect(screen.getByLabelText("Email Address").value).toBe("");
    });
});
