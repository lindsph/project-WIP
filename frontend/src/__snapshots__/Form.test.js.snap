// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> renders correctly with the default props 1`] = `
<body>
  <div>
    <div
      class="form-wrapper"
    >
      <h1
        class="form-heading"
      >
        Sign in
      </h1>
      <h2
        class="form-heading"
      >
        Already have an account? Sign in instead.
      </h2>
      <form>
        <div
          class="form-group"
        >
          <label
            for="email"
          >
            Email Address
          </label>
          <input
            aria-invalid="false"
            id="email"
            name="email"
            type="text"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="password"
          >
            Password
          </label>
          <input
            aria-invalid="false"
            id="password"
            name="password"
            type="text"
          />
        </div>
        <button
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>
</body>
`;
